### PR TITLE
Zhao dev: add an exp function for Tensor2_3D

### PR DESCRIPTION
**Pull request by Chongran-Zhao**
The algorithm can be found on page 749 of Computational Methods for Plasticity written by E. A. De Souza Neto, D. Perić and D. R. J. Owen..

### CODE REVIEW

File: src/System/Tensor2_3D.cpp [hunk 1]
```diff
@@ -596,4 +596,26 @@ Tensor2_3D Ten2::gen_zero()
...
+
+  do
+  {
+      nn += 1.0;
+      nn_fac *= nn;
+
+      const Tensor2_3D temp = input_pow * input;
+      input_pow = temp;
+      input_exp += ( 1.0/nn_fac ) * input_pow;
+
```
Comment (guanjiashen) [inline src/System/Tensor2_3D.cpp:615]: 缩进太多了。其他没什么问题。

**Review by guanjiashen — APPROVED** [review]

File: src/System/Tensor2_3D.cpp [hunk 2]
```diff
@@ -596,4 +596,26 @@ Tensor2_3D Ten2::gen_zero()
       0.0, 0.0, 0.0 );
 }
 
+Tensor2_3D Ten2::exp( const Tensor2_3D &input )
+{
+  double nn = 0.0;
+  long double nn_fac = 1.0;
```
Comment (ju-liu) [inline src/System/Tensor2_3D.cpp:602]: 是否有必要使用long在这里

Comment (Chongran-Zhao) [reply src/System/Tensor2_3D.cpp]: 没有必要，已经删除，double范围是-1.7E-308~1.7E308

File: src/System/Tensor2_3D.cpp (hunk 1 above)
Comment (Chongran-Zhao) [reply src/System/Tensor2_3D.cpp:615]: 已修改

File: src/System/Tensor2_3D.cpp [hunk 3]
```diff
@@ -596,4 +596,26 @@ Tensor2_3D Ten2::gen_zero()
...
+  Tensor2_3D input_pow = Ten2::gen_id();
+  Tensor2_3D input_exp = Ten2::gen_id();
+
+  do
+  {
+    nn += 1.0;
+    nn_fac *= nn;
+
+    const Tensor2_3D temp = input_pow * input;
+    input_pow = temp;
```
Comment (ju-liu) [inline src/System/Tensor2_3D.cpp:613]: another question: do we really need a temp tensor here. Can we just do input_pow = input_pow * input?

Comment (ju-liu) [reply src/System/Tensor2_3D.cpp:613]: I assume our * operator allows us to do so. @guanjiashen can you confirm?

Comment (guanjiashen) [reply src/System/Tensor2_3D.cpp:613]: My tests tell me we can use input_pow = input_pow  * input. In fact, the * operator only returns a new Tensor2_3D data.

Comment (Chongran-Zhao) [reply src/System/Tensor2_3D.cpp]: Yes, I have deleted the temp tensor.

File: src/System/Tensor2_3D.cpp [hunk 4]
```diff
@@ -596,4 +596,25 @@ Tensor2_3D Ten2::gen_zero()
...
+  Tensor2_3D input_pow = Ten2::gen_id();
+  Tensor2_3D input_exp = Ten2::gen_id();
+
+  do
+  {
+    nn += 1.0;
+    nn_fac *= nn;
+
+    input_pow = input_pow * input;
+    input_exp += ( 1.0/nn_fac ) * input_pow;
```
Comment (ju-liu) [inline src/System/Tensor2_3D.cpp:613]: 为什么不 input_pow / nn_fac 
这样只用一次除法，而不是除法 和 乘法。

Comment (guanjiashen) [reply src/System/Tensor2_3D.cpp:613]: Sir, you're right, but we do not have the operation Tensor2_3D / scalar yet.

Comment (Chongran-Zhao) [reply src/System/Tensor2_3D.cpp:613]: Tensor2_3D没有定义张量对标量的除法运算：
```bash
 error: invalid operands to binary expression ('Tensor2_3D' and 'double')
    input_exp += input_pow / nn_fac;
```

**Review by ju-liu — APPROVED** [review]

**Pull request merged**